Feature/sdl 0188 interior vehicle data resumption

AKalinich-Luxoft opened the pull request:
Implements #133 

This PR is **ready** for review.

### Note
This PR should be merged only after #394 as it based on implementation of another feature

### Testing Plan
Will be tested manually

[Manual instruction for SDL-0188 Interior Vehicle Data resumption.zip](https://github.com/smartdevicelink/sdl_hmi/files/5104497/Manual.instruction.for.SDL-0188.Interior.Vehicle.Data.resumption.zip)


### Summary
- Added GetIVD control form and logic
- Add custom response logic for GetIVD request

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

AKalinich-Luxoft commented:
@yang1070 please consider the following updates since our last code review:
1. Updated year in the copyright comments (https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/d809cb7e7ec0fddce53be6aeec603038aa7e619f)
2. Removed unrelated images (https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/a6d542d6de2f3b88c58c0303fed8cf3a1254d1af)
3. Add separate code customization for choice sets (https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/a2f8aeaf9cbce3356569a02cf8c748a71027c34f)
4. Add isSubscribed flag emulation (https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/0cc8a1d61b98a2912fcee4e55b6fd977f5cb3f17)

yang1070 commented:
@AKalinich-Luxoft please also update PR394 for SDL-190 regarding copyright year and remove unrelated images

KhrystynaDubovyk commented:
@yang1070, please approve this PR

AKalinich-Luxoft commented:
> @AKalinich-Luxoft please also update PR394 for SDL-190 regarding copyright year and remove unrelated images

@yang1070 current PR has been rebased on top of 0190 implementation to remove unrelated changes.

The following three commits have been moved to 0190 implementation:
1. Updated year in the copyright comments (6b7dcd1)
2. Removed unrelated images (21b7977)
3. Add separate code customization for choice sets (fa5f862)

Implementation of 0188 still contain the following commit since our last review:
1. Add isSubscribed flag emulation (7baa8a4)

ShobhitAd commented:
The UI elements in `RPCGetIVDControlConfigView` don't seem to be updated in certain cases when a `GetInteriorVehicleData` request is processed. For example, in the case where I: 
1) add two extra results(3 results total)
2) press the `Prev` button once(counter label reads 2/3)
3) send the GIVD request while keeping the view open
4) press the `Next` button(counter label reads 3/2). 

![Screenshot from 2020-08-27 14-06-43](https://user-images.githubusercontent.com/10703941/91480072-6794af00-e870-11ea-95ca-24e912fbb93f.png)

In this case, after the GIVD request is processed, I would expect the `Next` button to be disabled.

YarikMamykin commented on app/view/settings/RPCControls/RPCGetIVDControlConfigView.js:
```diff
@@ -0,0 +1,160 @@
+      'previousButton',
+      'nextButton',
+      'counterLabel',
+      'getIVDSelect',
```
@ShobhitAd excellent idea! Implemented in https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/be5866cf51e34e0ec5b743403ba9b89b0f9397b7

YarikMamykin commented on ffw/RCRPC.js:
```diff
@@ -250,26 +250,56 @@ FFW.RC = FFW.RPCObserver.create(
+              };
+
+              var data = SDL.RCModulesController.getInteriorVehicleData(request);
+                if (data) {
```
@ShobhitAd indentation fixed in https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/1441815a5201af465df43d7525b1cb33fd00da97

YarikMamykin commented on ffw/RCRPC.js:
```diff
@@ -250,26 +250,56 @@ FFW.RC = FFW.RPCObserver.create(
+                'result': {
+                    'code': resultStruct.code,
+                    'method': request.method,
+                    'isSubscribed': calculate_subscribed_value(resultStruct.subscribed),
```
@ShobhitAd accordingly to the API if `isSubscribed` param is undefined (not `true` and not `false`) we should not send it in response. Please take a look at
https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/f5df4cbe063118adf06e1cecf4b9621f5130ed39

YarikMamykin commented:
> The UI elements in `RPCGetIVDControlConfigView` don't seem to be updated in certain cases when a `GetInteriorVehicleData` request is processed. For example, in the case where I:
> 
>     1. add two extra results(3 results total)
> 
>     2. press the `Prev` button once(counter label states "2/3")
> 
>     3. send the GIVD request while keeping the view open(counter label still states "2/3")
> 
>     4. press the `Next` button(counter label states "3/2").
> 
> 
> ![Screenshot from 2020-08-27 14-06-43](https://user-images.githubusercontent.com/10703941/91480072-6794af00-e870-11ea-95ca-24e912fbb93f.png)
> 
> In this case, after the GIVD request is processed, I would expect the `Next` button to be disabled and the counter label to be update

@ShobhitAd this issue fixed in https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/a52e0b1e43a20b7b47c7cceaf0ea1c2b5b6ad3f1

YarikMamykin commented:
@ShobhitAd comments are processed. Please take a look on latest changes.

ShobhitAd commented:
> @ShobhitAd comments are processed. Please take a look on latest changes.

@YarikMamykin thanks for making the changes. Since this PR is dependent on the changes in https://github.com/smartdevicelink/sdl_hmi/pull/394, could you also merge in the latest changes from `feature/sdl_0190_resumption_data_error_handling`. 

There seem to be some conflicts at the moment in https://github.com/LuxoftSDL/sdl_hmi/compare/feature/sdl_0190_resumption_data_error_handling...LuxoftSDL:feature/sdl_0188_interior_vehicle_data_resumption

ShobhitAd commented:
> > The UI elements in `RPCGetIVDControlConfigView` don't seem to be updated in certain cases when a `GetInteriorVehicleData` request is processed. For example, in the case where I:
> > ```
> > 1. add two extra results(3 results total)
> > 
> > 2. press the `Prev` button once(counter label states "2/3")
> > 
> > 3. send the GIVD request while keeping the view open(counter label still states "2/3")
> > 
> > 4. press the `Next` button(counter label states "3/2").
> > ```
> > 
> > 
> > ![Screenshot from 2020-08-27 14-06-43](https://user-images.githubusercontent.com/10703941/91480072-6794af00-e870-11ea-95ca-24e912fbb93f.png)
> > In this case, after the GIVD request is processed, I would expect the `Next` button to be disabled and the counter label to be update
> 
> @ShobhitAd this issue fixed in [3255157](https://github.com/smartdevicelink/sdl_hmi/commit/3255157559529a6841cb5a237155492d18b2b7f5)

@YarikMamykin Even though the counter label is updated correctly to "2/2" after step 3, the next button is still active so I'm hitting the same issue "3/2" in step 4

YarikMamykin commented:
> > > The UI elements in `RPCGetIVDControlConfigView` don't seem to be updated in certain cases when a `GetInteriorVehicleData` request is processed. For example, in the case where I:
> > > ```
> > > 1. add two extra results(3 results total)
> > > 
> > > 2. press the `Prev` button once(counter label states "2/3")
> > > 
> > > 3. send the GIVD request while keeping the view open(counter label still states "2/3")
> > > 
> > > 4. press the `Next` button(counter label states "3/2").
> > > ```
> > > 
> > > 
> > > ![Screenshot from 2020-08-27 14-06-43](https://user-images.githubusercontent.com/10703941/91480072-6794af00-e870-11ea-95ca-24e912fbb93f.png)
> > > In this case, after the GIVD request is processed, I would expect the `Next` button to be disabled and the counter label to be update
> > 
> > 
> > @ShobhitAd this issue fixed in [3255157](https://github.com/smartdevicelink/sdl_hmi/commit/3255157559529a6841cb5a237155492d18b2b7f5)
> 
> @YarikMamykin Even though the counter label is updated correctly to "2/2" after step 3, the next button is still active so I'm hitting the same issue "3/2" in step 4

@ShobhitAd okay, I see there is some "bad magic" in Ember.js framework event handling. Added small adjustment for this in https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/3499985d1129ae1a5379b406fc905c990020e3c9

YarikMamykin commented:
> > @ShobhitAd comments are processed. Please take a look on latest changes.
> 
> @YarikMamykin thanks for making the changes. Since this PR is dependent on the changes in #394, could you also merge in the latest changes from `feature/sdl_0190_resumption_data_error_handling`.
> 
> There seem to be some conflicts at the moment in [LuxoftSDL/sdl_hmi@feature/sdl_0190_resumption_data_error_handling...LuxoftSDL:feature/sdl_0188_interior_vehicle_data_resumption](https://github.com/LuxoftSDL/sdl_hmi/compare/feature/sdl_0190_resumption_data_error_handling...LuxoftSDL:feature/sdl_0188_interior_vehicle_data_resumption)

@ShobhitAd backmerge of `feature/sdl_0190_resumption_data_error_handling` into `feature/sdl_0188_interior_vehicle_data_resumption` is done in https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/52310d67faf97f5c03994fbda8fa943072fcd827

ShobhitAd commented:
> > > > The UI elements in `RPCGetIVDControlConfigView` don't seem to be updated in certain cases when a `GetInteriorVehicleData` request is processed. For example, in the case where I:
> > > > ```
> > > > 1. add two extra results(3 results total)
> > > > 
> > > > 2. press the `Prev` button once(counter label states "2/3")
> > > > 
> > > > 3. send the GIVD request while keeping the view open(counter label still states "2/3")
> > > > 
> > > > 4. press the `Next` button(counter label states "3/2").
> > > > ```
> > > > 
> > > > 
> > > > ![Screenshot from 2020-08-27 14-06-43](https://user-images.githubusercontent.com/10703941/91480072-6794af00-e870-11ea-95ca-24e912fbb93f.png)
> > > > In this case, after the GIVD request is processed, I would expect the `Next` button to be disabled and the counter label to be update
> > > 
> > > 
> > > @ShobhitAd this issue fixed in [3255157](https://github.com/smartdevicelink/sdl_hmi/commit/3255157559529a6841cb5a237155492d18b2b7f5)
> > 
> > 
> > @YarikMamykin Even though the counter label is updated correctly to "2/2" after step 3, the next button is still active so I'm hitting the same issue "3/2" in step 4
> 
> @ShobhitAd okay, I see there is some "bad magic" in Ember.js framework event handling. Added small adjustment for this in [3499985](https://github.com/smartdevicelink/sdl_hmi/commit/3499985d1129ae1a5379b406fc905c990020e3c9)

@YarikMamykin thanks, that seems to have fixed the issue

AKalinich-Luxoft commented:
One more backmerge after 0190 feature merge into develop was done in https://github.com/smartdevicelink/sdl_hmi/pull/395/commits/3f37524cc1bafeef22751a3bb96d09889f4ef0cb
